tests/integration2/pilot/sidecar_api_test.go

```diff
@@ -0,0 +1,134 @@
+package pilot
+
+import (
+	"fmt"
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/api/components"
+	"istio.io/istio/pkg/test/framework/api/descriptors"
+	"istio.io/istio/pkg/test/framework/api/ids"
+	"istio.io/istio/pkg/test/framework/api/lifecycle"
+	"istio.io/istio/pkg/test/framework/runtime/components/environment/native"
+	"istio.io/istio/pkg/test/scopes"
+	"istio.io/istio/pkg/test/util/structpath"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+func TestSidecarListeners(t *testing.T) {
+	// Call Requires to explicitly initialize dependencies that the test needs.
+	ctx := framework.GetContext(t)
+	scopes.Framework.SetOutputLevel(log.DebugLevel)
+
+	// TODO: Limit to Native environment until the Kubernetes environment is supported in the Galley
+	// component
+	ctx.RequireOrSkip(t, lifecycle.Test, &descriptors.NativeEnvironment)
+	ctx.RequireOrFail(t, lifecycle.Test, &ids.Galley)
+	ctx.RequireOrFail(t, lifecycle.Test, &ids.Pilot)
+
+	// Load the canonical dataset into Galley and by implication Pilot
+	gal := components.GetGalley(ctx, t)
+	env, err := native.GetEnvironment(ctx)
+	components.GetMixer(ctx, t)
+	s := strings.Split(env.Mesh.MixerCheckServer, ":")[1]
+	mixerPort, _ := strconv.ParseUint(s, 10, 32)
+
+	path, err := filepath.Abs("../../testdata/config")
+	if err != nil {
+		t.Fatalf("No such directory: %v", err)
+	}
+	gal.ApplyConfigDir(path)
+
+	pilot := components.GetPilot(ctx, t)
+
+	// Simulate proxy identity of egress gateway ...
+	nodeId := &model.Proxy{
+		ClusterID:       "integration-test",
+		Type:            model.SidecarProxy,
+		IPAddresses:     []string {"10.2.0.1"},
+		ID:              "app3.testns",
+		DNSDomain:       "testns.cluster.local",
+	}
+
+	// ... and get listeners from Pilot for that proxy
+	req := &xdsapi.DiscoveryRequest{
+		Node: &core.Node{
+			Id:  nodeId.ServiceNode(),
+		},
+		TypeUrl: "type.googleapis.com/envoy.api.v2.Listener",
+	}
+	resp, err := pilot.CallDiscovery(req)
+	if err != nil {
+		t.Fatalf("discovery failed: %v", err)
+	}
+
+	response, _ := structpath.New(resp)
+	t.Run("validate-mixer-listener", func(t *testing.T) {
```

Comment:
General comment here: I think this is a good approach. a much better improvement from having a 10k line golden file in the repo and comparing it. Would it be easier in the long run, if you compared the XDS object returned with your XDS object template? For example,

``` 
expectedResponse = core.Listener{ 
   name: "foobar", 
   listenerFilters: toStruct({
           name: http_connection_manager, 
           config: { } 
     })
  }

```

when you compare the expectedResponse with an actual response, like this:

```
 Equals(expectedResponse, actualResponse)
   if expectedResponse.field == nil {skip // i.e. we dont care about it }
   else {
    if expectedResponse.field is primitiveType, return expectedResponse.field == actualResponse.field
    else Equals(expectedResponse.field, actualResponse.field)
```

The equals code above achieves the same purpose as the MatchString/bool/etc. you have

The advantage of comparing protos or proto objects is that its easier to grok at what came out of pilot, and compare it to what we wrote in the code.

Reply:
+1 ... the if/then/else is harder to read and will be harder to reproduce. I expect we'll have a lot of tests that are validating Envoy config going forward.

Reply:
The potential issue here is nil having dual value, I think (we expect nil v.s. we don't care).

Reply:
YEs. but that is left to the test writer. If I want to validate that the field is nil, I can supply a flag stating so. if I want to indicate that nil implies 'do not care', we can use a flag for the same.

Reply:
A few points on this topic 

- Most test contexts won't be testing the entire structure of a resource so using goldens or structs are usually sub-optimal for these classes of tests as they introduce a lot of noise and the source of a difference is not  easily annotated in the test. What I expect is that there is a baseline of tests that are very thorough across the API domain and then as features are added a suite of smaller tests are added that check a more restricted set of points of behavior in the API

- Negative tests are hard to do with struct matches. E.g. the test that checks that original_dest is not enabled/set

- Supplying flags to a utility method for nils would only cover part of the problem, a more painful use-case is array membership. Trying to mimic positional array matches in long lists without filling out the 'padding' around the entry you're trying to test can hurt a lot and becomes brittle when order is not guaranteed. The use of match queries in jsonpath solves this pretty well. Struct matches are fine for things that are concrete and not that deeply nested. (Note that structpath supports a FindStruct to enable this if you want to use that form in conjunction. I could also add a MatchStruct so these can be co-mingled)

-

Reply:
PTAL at the updates, switched to the fluent style per Oz's suggestion

Added an example of struct comparison within the framework. Worth nothing how the pervasive use of one-of's impacts readability

Reply:
@rshriram re: any. Switch to `proto.Any` is gated by ISTIO_PROXY_VERSION node metadata. You should be able to convert Any to Struct universally in the test if necessary. The plan is to drop Struct since it is deprecated by envoy.

Reply:
This is stable in the tests but I believe the proto->json serialization will expand Anys based on their type information if its available within the binary. 

Ill try this out in a follow up